File: packages/react-native-audio-api/common/cpp/audioapi/core/effects/PannerNode.cpp
#include <audioapi/core/AudioListener.h>
#include <audioapi/core/BaseAudioContext.h>
#include <audioapi/core/effects/PannerNode.h>
#include <audioapi/core/effects/PannerSpatialization.h>
#include <audioapi/core/utils/Constants.h>
#include <audioapi/types/NodeOptions.h>
#include <audioapi/utils/AudioArray.hpp>

#include <cmath>
#include <memory>

namespace audioapi {

namespace {

using panner::DEG_180;
using panner::DEG_90;
using panner::Vec3;

std::shared_ptr<AudioParam> createPannerParam(
    float defaultValue,
    const std::shared_ptr<BaseAudioContext> &context) {
  return std::make_shared<AudioParam>(
      defaultValue, MOST_NEGATIVE_SINGLE_FLOAT, MOST_POSITIVE_SINGLE_FLOAT, context);
}

} // namespace

PannerNode::PannerNode(
    const std::shared_ptr<BaseAudioContext> &context,
    AudioListener *listener,
    const PannerOptions &options)
    : AudioNode(context, options),
      listener_(listener),
      positionXParam_(createPannerParam(options.positionX, context)),
      positionYParam_(createPannerParam(options.positionY, context)),
      positionZParam_(createPannerParam(options.positionZ, context)),
      orientationXParam_(createPannerParam(options.orientationX, context)),
      orientationYParam_(createPannerParam(options.orientationY, context)),
      orientationZParam_(createPannerParam(options.orientationZ, context)),
      panningModel_(options.panningModel),
      distanceModel_(options.distanceModel),
      refDistance_(options.refDistance),
      maxDistance_(options.maxDistance),
      rolloffFactor_(options.rolloffFactor),
      coneInnerAngle_(options.coneInnerAngle),
      coneOuterAngle_(options.coneOuterAngle),
      coneOuterGain_(options.coneOuterGain),
      outputBuffer_(
          std::make_shared<DSPAudioBuffer>(
              RENDER_QUANTUM_SIZE,
              channelCount_,
              context->getSampleRate())) {}

void PannerNode::processNode(int framesToProcess) {
  std::shared_ptr<BaseAudioContext> context = context_.lock();
  if (context == nullptr || audioBuffer_ == nullptr || listener_ == nullptr) {
    outputBuffer_->zero();
    return;
  }

  const double time = context->getCurrentTime();
  const bool monoInput = audioBuffer_->getNumberOfChannels() == 1;

  listener_->processForQuantum(framesToProcess, time, context->getCurrentSampleFrame());

  const auto posX =
      positionXParam_->processARateParam(framesToProcess, time)->getChannel(0)->span();
  const auto posY =
      positionYParam_->processARateParam(framesToProcess, time)->getChannel(0)->span();
  const auto posZ =
      positionZParam_->processARateParam(framesToProcess, time)->getChannel(0)->span();
  const auto orientX =
      orientationXParam_->processARateParam(framesToProcess, time)->getChannel(0)->span();
  const auto orientY =
      orientationYParam_->processARateParam(framesToProcess, time)->getChannel(0)->span();
  const auto orientZ =
      orientationZParam_->processARateParam(framesToProcess, time)->getChannel(0)->span();

  const auto listenerPosX = listener_->positionXValues();
  const auto listenerPosY = listener_->positionYValues();
  const auto listenerPosZ = listener_->positionZValues();
  const auto listenerForwardX = listener_->forwardXValues();
  const auto listenerForwardY = listener_->forwardYValues();
  const auto listenerForwardZ = listener_->forwardZValues();
  const auto listenerUpX = listener_->upXValues();
  const auto listenerUpY = listener_->upYValues();
  const auto listenerUpZ = listener_->upZValues();

  auto outputLeft = outputBuffer_->getChannelByType(AudioBuffer::ChannelLeft)->span();
  auto outputRight = outputBuffer_->getChannelByType(AudioBuffer::ChannelRight)->span();

  auto inputLeftSpan = monoInput ? audioBuffer_->getChannelByType(AudioBuffer::ChannelMono)->span()
                                 : audioBuffer_->getChannelByType(AudioBuffer::ChannelLeft)->span();
  auto inputRightSpan =
      monoInput ? inputLeftSpan : audioBuffer_->getChannelByType(AudioBuffer::ChannelRight)->span();

  // Equal-power is the only supported panning model.
  (void)panningModel_;

  for (int i = 0; i < framesToProcess; ++i) {
    const auto idx = static_cast<size_t>(i);
    const Vec3 sourcePosition{.x = posX[idx], .y = posY[idx], .z = posZ[idx]};
    const Vec3 sourceOrientation{.x = orientX[idx], .y = orientY[idx], .z = orientZ[idx]};
    const Vec3 listenerPosition{
        .x = listenerPosX[idx], .y = listenerPosY[idx], .z = listenerPosZ[idx]};
    const Vec3 listenerForward{
        .x = listenerForwardX[idx], .y = listenerForwardY[idx], .z = listenerForwardZ[idx]};
    const Vec3 listenerUp{.x = listenerUpX[idx], .y = listenerUpY[idx], .z = listenerUpZ[idx]};

    double azimuth =
        panner::computeAzimuth(sourcePosition, listenerPosition, listenerForward, listenerUp);

    azimuth = panner::clampAzimuth(azimuth);
    azimuth = panner::wrapAzimuth(azimuth);

    double x = 0.0;
    if (monoInput) {
      x = (azimuth + DEG_90) / DEG_180;
    } else if (azimuth <= 0.0) {
      x = (azimuth + DEG_90) / DEG_90;
    } else {
      x = azimuth / DEG_90;
    }

    const double gainL = std::cos(x * panner::PI / 2.0);
    const double gainR = std::sin(x * panner::PI / 2.0);

    const double inputL = inputLeftSpan[idx];
    const double inputR = inputRightSpan[idx];

    double panLeft = 0.0;
    double panRight = 0.0;

    if (monoInput) {
      panLeft = inputL * gainL;
      panRight = inputL * gainR;
    } else if (azimuth <= 0.0) {
      panLeft = inputL + inputR * gainL;
      panRight = inputR * gainR;
    } else {
      panLeft = inputL * gainL;
      panRight = inputR + inputL * gainR;
    }

    const double distance = panner::computeDistance(sourcePosition, listenerPosition);
    const double distanceGain = panner::computeDistanceGain(
        distanceModel_, distance, refDistance_, maxDistance_, rolloffFactor_);
    const double coneGain = panner::computeConeGain(
        sourcePosition,
        listenerPosition,
        sourceOrientation,
        coneInnerAngle_,
        coneOuterAngle_,
        coneOuterGain_);
    const double totalGain = coneGain * distanceGain;

    outputLeft[idx] = static_cast<float>(totalGain * panLeft);
    outputRight[idx] = static_cast<float>(totalGain * panRight);
  }
}

} // namespace audioapi

File: packages/react-native-audio-api/src/core/PannerNode.ts
import { IPannerNode } from '../jsi-interfaces';
import {
  ChannelCountMode,
  DistanceModelType,
  PannerOptions,
  PanningModelType,
} from '../types';
import { InvalidStateError, NotSupportedError, RangeError } from '../errors';
import {
  PannerOptionsValidator,
  validatePannerChannelCount,
  validatePannerChannelCountMode,
} from '../utils/validation';
import { assertFloat32Representable } from '../utils/float32';
import AudioNode from './AudioNode';
import AudioParam from './AudioParam';
import type BaseAudioContext from './BaseAudioContext';

export default class PannerNode extends AudioNode {
  readonly positionX: AudioParam;
  readonly positionY: AudioParam;
  readonly positionZ: AudioParam;
  readonly orientationX: AudioParam;
  readonly orientationY: AudioParam;
  readonly orientationZ: AudioParam;

  constructor(context: BaseAudioContext, options?: PannerOptions) {
    PannerOptionsValidator.validate(options);
    const panner: IPannerNode = context.context.createPanner(options || {});
    super(context, panner);
    this.positionX = new AudioParam(panner.positionX, context, this);
    this.positionY = new AudioParam(panner.positionY, context, this);
    this.positionZ = new AudioParam(panner.positionZ, context, this);
    this.orientationX = new AudioParam(panner.orientationX, context, this);
    this.orientationY = new AudioParam(panner.orientationY, context, this);
    this.orientationZ = new AudioParam(panner.orientationZ, context, this);
  }

  public get panningModel(): PanningModelType {
    return (this.node as IPannerNode).panningModel;
  }

  public set panningModel(value: PanningModelType) {
    if (value === 'HRTF') {
      throw new NotSupportedError(
        "panningModel 'HRTF' is not supported yet; use 'equalpower'"
      );
    }
    (this.node as IPannerNode).panningModel = value;
  }

  public override get channelCount(): number {
    return this.node.channelCount;
  }

  public override set channelCount(value: number) {
    validatePannerChannelCount(value);
    this.node.channelCount = value;
  }

  public override get channelCountMode(): ChannelCountMode {
    return this.node.channelCountMode;
  }

  public override set channelCountMode(value: ChannelCountMode) {
    validatePannerChannelCountMode(value);
    this.node.channelCountMode = value;
  }

  public get distanceModel(): DistanceModelType {
    return (this.node as IPannerNode).distanceModel;
  }

  public set distanceModel(value: DistanceModelType) {
    (this.node as IPannerNode).distanceModel = value;
  }

  public get refDistance(): number {
    return (this.node as IPannerNode).refDistance;
  }

  public set refDistance(value: number) {
    if (value < 0) {
      throw new RangeError('refDistance must be non-negative');
    }
    (this.node as IPannerNode).refDistance = value;
  }

  public get maxDistance(): number {
    return (this.node as IPannerNode).maxDistance;
  }

  public set maxDistance(value: number) {
    if (value <= 0) {
      throw new RangeError('maxDistance must be positive');
    }
    (this.node as IPannerNode).maxDistance = value;
  }

  public get rolloffFactor(): number {
    return (this.node as IPannerNode).rolloffFactor;
  }

  public set rolloffFactor(value: number) {
    if (value < 0) {
      throw new RangeError('rolloffFactor must be non-negative');
    }
    (this.node as IPannerNode).rolloffFactor = value;
  }

  public get coneInnerAngle(): number {
    return (this.node as IPannerNode).coneInnerAngle;
  }

  public set coneInnerAngle(value: number) {
    (this.node as IPannerNode).coneInnerAngle = value;
  }

  public get coneOuterAngle(): number {
    return (this.node as IPannerNode).coneOuterAngle;
  }

  public set coneOuterAngle(value: number) {
    (this.node as IPannerNode).coneOuterAngle = value;
  }

  public get coneOuterGain(): number {
    return (this.node as IPannerNode).coneOuterGain;
  }

  public set coneOuterGain(value: number) {
    if (value < 0 || value > 1) {
      throw new InvalidStateError('coneOuterGain must be between 0 and 1');
    }
    (this.node as IPannerNode).coneOuterGain = value;
  }

  public setPosition(x: number, y: number, z: number): void {
    assertFloat32Representable(x);
    assertFloat32Representable(y);
    assertFloat32Representable(z);
    this.positionX.value = x;
    this.positionY.value = y;
    this.positionZ.value = z;
  }

  public setOrientation(x: number, y: number, z: number): void {
    assertFloat32Representable(x);
    assertFloat32Representable(y);
    assertFloat32Representable(z);
    this.orientationX.value = x;
    this.orientationY.value = y;
    this.orientationZ.value = z;
  }
}
